lib/active_merchant/billing/gateways/square.rb

```diff
@@ -0,0 +1,375 @@
+        end
+
+        post = {}.merge(options.slice(:buyer_email_address, :delay_capture, :note, :reference_id))
+        add_idempotency_key(post, options)
+        add_amount(post, money, options)
+        add_address(post, options)
+        post[:reference_id] = options[:order_id] if options[:order_id]
+        post[:note] = options[:description] if options[:description]
+
+        MultiResponse.run do |r|
+          if options[:customer] && card_nonce
+            # Since customer was passed in, create customer (if needed) and store card (always in here).
+            options[:customer][:customer_id] = options[:customer][:id] if options[:customer][:id] # To make store() happy.
+            r.process { store(card_nonce, options[:customer]) }
+
+            # If we just created a customer.
+            if options[:customer][:id].nil?
+              options[:customer][:id] = r.responses.first.params['customer']['id']
+            end
+
+            # We always stored a card, so grab it.
+            options[:customer][:card_id] = r.responses.last.params['card']['id']
+
+            card_nonce = nil # empty the card_nonce, since we now have the card on file
+            
+            # Invariant: we have a customer and a linked card, and our options hash is correct.
+          end
+
+          add_payment(post, card_nonce, options)
+          r.process { commit(:post, "locations/#{@location_id}/transactions", post) }
+        end
+      end
+
+      # Authorize for Square uses the Charge with delay_capture = true option. 
+      # https://docs.connect.squareup.com/api/connect/v2/#endpoint-charge
+      # Same as with `purchase`, pass nil for `card_nonce` if using a customer's
+      # stored card on file.
+      # 
+      # See purchase for more details for calling this.
+      def authorize(money, card_nonce, options={})
+        options[:delay_capture] = true
+        purchase(money, card_nonce, options)
+      end
+
+      # Capture is only used if you did an Authorize, (creating a delayed capture).
+      # https://docs.connect.squareup.com/api/connect/v2/#endpoint-capturetransaction
+      # Both `money` and `options` are unused. Only a full capture is supported.
+      def capture(ignored_money, txn_id, ignored_options={})
+        raise ArgumentError('txn_id required') if txn_id.nil?
+        commit(:post, "locations/#{CGI.escape(@location_id)}/transactions/#{CGI.escape(txn_id)}/capture")
+      end
+
+      # Refund refunds a previously Charged transaction. 
+      # https://docs.connect.squareup.com/api/connect/v2/#endpoint-createrefund
+      # Options require: `tender_id`, and permit `idempotency_key`, `reason`.
+      def refund(money, txn_id, options={})
+        raise ArgumentError('txn_id required') if txn_id.nil?
+        raise ArgumentError('money required') if money.nil?
+        requires!(options, :tender_id)
+        
+        post = {}.merge(options.slice(:tender_id, :reason))
```

Comment:
Same here. `Hash#slice` (which is Rails specific by the way) returns a new Hash.